### DISCUSSION TITLE
Versioning and automating app deployments for internal builds

### DISCUSSION THREAD

**Discussion by ColtonIdle** [MobileNativeFoundation/discussions#165] [General]
Just wanted to start a discussion on how versioning and automating app deployments 

Essentially we're facing some analysis paralysis:

Do we

1. Manually tag a release as new version 1.0.1, then the same dev press a button to kickoff CI to look for the last tag. and create a build. send it to firebase app distrubition. send a slack notif. etc
or
2. have a CI job that a dev invokes, and it adds a tag itself and then creates a build to shipoff to firebase app dist.

What we're getting caught up on is what "operation" actually signifies a new release of the app. Is it the presence of this new app on firebase app distribution or does the tag in the repo signify a release (even though the release may not be out yet)

It of course would be superrrr nice to have it automated. which is why optino 1 kinda feels "wrong" because the user has to manually tag and bump versions. but maybe thats just a fact of life.

the con of option 2 is that the CI creates a build and will also make the PR for the version bumps, but it technically happens after the release is sent.

We have option 2 implemented. and its nice since its all automated. But i think option 1 is actually better?

**Comment by ialimz** [comment]
In the case of implementing option 2, would CI itself know which version number it should create? Or do we have to pass the version number as an argument(or maybe get it from info.plist)?

We use option 1 in our project. here are some of the reasons I have to prefer option 1:
1- We can write custom release notes during tag creation
2- We can trigger CI actions on tags we just created

**Reply by ColtonIdle** [reply]
Yeah. The big thing for me is that in option 2 we have the CI put together the release notes automatically, but we want to commit this back to the codebase but we can't really have the CI create a PR with the release note list. This might be easier to automate in a local script that a user kicks off (versions the app and release note creation, and tags it) then the CI just builds that tag.

**Comment by eliperkins** [comment]
My teammate @taki-on [wrote up a fantastic blog post about how we're doing with this GitHub for iOS](https://github.blog/2022-01-12-how-we-ship-github-mobile-every-week/). I'd highly recommend giving it a read through on how we're approaching shipping iOS builds to the App Store weekly, especially with regards to automation.

Some notes to answer some of the thoughts you pose here:

> What we're getting caught up on is what "operation" actually signifies a new release of the app. Is it the presence of this new app on firebase app distribution or does the tag in the repo signify a release (even though the release may not be out yet)

I think tags for pre-release stuff is totally fine! There's nothing wrong with a tag indicating a version that _isn't_ yet out on the App Store, but might be out on TestFlight/Firebase/whatever your beta app distro method is. One trick we use is to create tags with both the marketing versions (e.g., the `CFBundleShortVersionString` in the Info.plist) _and_ the build number (e.g.,the `CFBundleVersion` in the Info.plist). As such, we end up with tags for daily builds (e.g., `v1.20.0-10001` for Monday's build, `v1.20.0-10002` for Tuesday's build) which get marked as pre-release versions, and then when we publish a build to the App Store, that tag gets promoted to the current stable release. That way, when we publish a release on the App Store we can point to a specific set of builds that we tested during the week and put the App Store release notes on the stable tag + release.

It's worth noting that the `CFBundleShortVersionString` doesn't need to increment for _every_ TestFlight build. Only `CFBundleVersion` needs to be incremented for each build. This allows us to have multiple builds of the same _marketing_ version, keeping communication about "what version will `FEATURE_XYZ` be released in?" much easier to reason about ("it will be in v1.51.0, getting released next Tuesday").

**Reply by ColtonIdle** [reply]
Thanks. This was super helpful!
